### PR TITLE
DEV: Centralize user updates to a single MessageBus channel.

### PR DESCRIPTION
**Pull request by tgxworld**
Introduces an interface to publish user updates on the server side and
helps to reduce the growing number of subscriptions on the client side.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/initializers/subscribe-user-notifications.js [hunk 1]
```diff
@@ -108,17 +108,19 @@ export default {
...
-
-      bus.subscribe(`/user-status/${user.id}`, (data) => {
-        user.set("status", data);
-        appEvents.trigger("user-status:changed");
+      bus.subscribe(`/user-updates/${user.id}`, (data) => {
+        switch (data.type) {
+          case "drafts":
+            user.updateDraftProperties(data.payload);
+            break;
+          case "do-not-disturb":
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/initializers/subscribe-user-notifications.js:116]: This is kind of silly, but can we make these snake_case instead? I find kebab-case-so-annoying-to-type, also we already use snake_case in other messagebus type cases like this :) (e.g. https://github.com/discourse/discourse/blob/e245839c3c9f58e60f643b940e1cc0240735e94f/app/assets/javascripts/discourse/app/controllers/topic.js#L1694-L1697)

File: app/models/user.rb [hunk 2]
```diff
@@ -657,16 +657,38 @@ def publish_notifications_state
     MessageBus.publish("/notification/#{id}", payload, user_ids: [id])
   end
 
+  PUBLISH_USER_STATUS_TYPE = "user-status"
```
Comment (martin-brennan) [inline app/models/user.rb:660]: See other comment about snake_case

**Review by martin-brennan — APPROVED** [review]
LGTM apart from some naming nits :+1:

File: app/assets/javascripts/discourse/app/initializers/subscribe-user-notifications.js (hunk 1 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/initializers/subscribe-user-notifications.js:116]: Also see https://github.com/discourse/discourse-chat/blob/9c646ebd1b46c10be29ff0bb67027d6fd2c6121c/assets/javascripts/discourse/components/chat-live-pane.js#L646-L646

**Comment by tgxworld** [comment]
Thank you for reviewing @martin-brennan

**Comment by AndrewPrigorshnev** [comment]
@tgxworld I wonder what was the aim of this change? What does it give us? Is it because of some performance concern?

In my opinion, this change adds a pretty serious problem in terms of readability and maintainability. Instead of several simple event handlers that's easy to read and understand, it creates one big handler with switch case inside. Also, it adds a `type` field to payload, which wasn't necessary before when names of channels were speaking for themselves.

Lots of small, simple functions is better than one big one that does everything.

Also, in general, it's a common pattern for messaging systems to have separate channels and handlers for separate events.
Handling every event in its separate handler is better approach, it improves quality of code, decreases mental load, and in this case there is no risk to accidentally break let's say drafts when fixing user status.

**Comment by tgxworld** [comment]
> What does it give us? Was that because of some performance concern?

This change reduces the number of subscriptions that is required on the client side. Although I have to be honest that I have not benchmark-ed this before and is something that was mentioned to me before. You can see how our messageBus subscription for a topic follows the same pattern: https://github.com/discourse/discourse/blob/f723b4c322e2d4f31d3be159d45fa08377c6d327/app/assets/javascripts/discourse/app/controllers/topic.js#L1620-L1687

> Also, it adds a type field to payload, which wasn't necessary before when names of channels were speaking for themselves.

Yea this is a trade-off here. If we're optimising for less channels to subscribe to, we need to slightly increase the size of the payload in order to be able to differentiate the messages.  

> this change adds a pretty serious problem in terms of readability and maintainability. Instead of several simple event handlers that's easy to read and understand, it creates one big handler with switch case inside.

I agree that reading it not as straight forward as having one handler per message type but I'm optimising for lesser channel subscriptions at the slight expense of readability. 

Maintainability wise, I wouldn't say it has suffered because the handler is meant to be light on application logic. If a handler is complex, the logic would likely be extracted into an appropriate class.

**Comment by AndrewPrigorshnev** [comment]
> You can see how our messageBus subscription for a topic follows the same pattern:

Frankly, I think that place is problematic as well, I would love if one day we split that method into separate handlers.

>  but I'm optimising for lesser channel subscriptions at the slight expense of readability.

But we don't have measurements. We don't know if it makes everything any faster. Without measurements, we cannot even be sure that this don't make things _slower_ in some unobvious way.

We can think about it in this way: we can compare to Ember. If we had just a few components instead of dozens of them, everything would probably work faster. But we never remove components to "reduce the growing number of components". The same is with event handlers.

Also, my gut feeling is that the cost we pay for having separate channels for the user status events, the DnD events and the  draft events should be negligible.

**Pull request merged**